### PR TITLE
storage,config: check size of lock cf when considering large CFs

### PR DESCRIPTION
**Pull request by nrc**
cc #7624, replaces #7417

### What problem does this PR solve?

When deciding to split regions, TiKV does not take into account the lock CF.

### What is changed and how it works?

Add lock CF to the CFs which are checked for size.

### Release note

No release note

PTAL @youjiali1995 @zhangjinpeng1987

### CODE REVIEW

**Comment by youjiali1995** [comment]
I glanced through the codes about region split:
* There is a worker to check whether a region needs to split and which the key is the split key. Every peer registers a tick to trigger the check periodically.
* There are several `Checker`s to check whether a region needs to split. For example:
  * [`KeysCheckObserver`](https://github.com/tikv/tikv/blob/master/components/raftstore/src/coprocessor/split_check/keys.rs): If the number of keys in the region is over the threshold, the region needs to split. I find it only takes into account `CF_WRITE`.
https://github.com/tikv/tikv/blob/782bb0203c3e39b2517bbb5ac0ceeef3dd76e60d/components/raftstore/src/coprocessor/split_check/keys.rs#L186
  * [`SizeCheckObserver`](https://github.com/tikv/tikv/blob/master/components/raftstore/src/coprocessor/split_check/size.rs): If the size of the region is over the threshold, the region needs to split. And the size contains `CF_LOCK` with your patch.
https://github.com/tikv/tikv/blob/782bb0203c3e39b2517bbb5ac0ceeef3dd76e60d/components/raftstore/src/coprocessor/split_check/size.rs#L207-L209

I think you can add tests in these checkers file. @nrc

**Comment by youjiali1995** [comment]
@zhangjinpeng1987 Should we take into account `CF_LOCK` in keys check?
https://github.com/tikv/tikv/blob/782bb0203c3e39b2517bbb5ac0ceeef3dd76e60d/components/raftstore/src/coprocessor/split_check/keys.rs#L186

And if we add table property to `CF_LOCK`, will it cause compatibility problems?

**Comment by BusyJay** [comment]
Now that large transaction is supported, I think it's reasonable to also considering lock cf. But remember that lock CF's key doesn't have TS suffix, which should be carefully processed as existing logic assumes it always exist.

**Comment by nrc** [comment]
> remember that lock CF's key doesn't have TS suffix

I've audited all uses of `LARGE_CFS` and as far as I can see, there is no use which assumes a TS encoding.

It is used in five files, often for testing where it is used as an arg to `rocks::util::new_engine_opt`. For the non-testing uses, it is used in `get_region_approximate_size` in components/raftstore/src/coprocessor/split_check/size.rs and in `scan_split_keys` in components/raftstore/src/store/worker/split_check.rs. In both cases the key is just treated as raw bytes, it is never assumed to have an encoding.

**Comment by nrc** [comment]
Looking at `get_region_approximate_keys_cf`, it is used in `get_region_approximate_keys` where it only looks at `WRITE_CF` (presumably because the number of keys must always be >= `DEFAULT_CF`). It is only used directly in `divide_db` in src/server/debug.rs. That uses write and default CFs and compares to find the larger one (I'm not sure why it doesn't just use write CF? Maybe in case of raw API, but in that case I would think `get_region_approximate_keys` should also look at default CF too).

AFAICT `get_region_approximate_keys_cf` does not assume a TS, so the only change necessary might be adding a lock CF check in `divide_db`, but that seems redundant to me.

**Comment by nrc** [comment]
> seems we should consider lock cf here too

I already have another commit ready :-) Will update soon.

**Comment by nrc** [comment]
@youjiali1995 I've added a commit so I think that the code is OK now, I haven't written a test because I am not really sure *what* to test. Also, I've been trying to benchmark, so far without any good results :-(

**Comment by youjiali1995** [comment]
> @youjiali1995 I've added a commit so I think that the code is OK now, I haven't written a test because I am not really sure _what_ to test. Also, I've been trying to benchmark, so far without any good results :-(

OK. I will take a deep look later and I'm not familiar with it too. 😂

**Comment by youjiali1995** [comment]
I found it only truncates timestamp for row key:
https://github.com/tikv/tikv/blob/45201a65dc4222444681dbbc5957860b5e21c981/components/raftstore/src/coprocessor/split_observer.rs#L36-L42

Why not do the same thing for index key? Will it result in different versions of an index key in more than one regions? @Connor1996

**Comment by youjiali1995** [comment]
I think we need to change here too:
https://github.com/tikv/tikv/blob/45201a65dc4222444681dbbc5957860b5e21c981/components/raftstore/src/coprocessor/split_check/keys.rs#L179-L186

Now it only considers the number of keys in write cf. I think we should add lock cf too. @nrc

**Comment by nrc** [comment]
> Now it only considers the number of keys in write cf. I think we should add lock cf too

@youjiali1995 yes, I wondered about that (https://github.com/tikv/tikv/pull/7676#issuecomment-620910114). I don't know why we don't check the default CF too?

**Comment by youjiali1995** [comment]
> > Now it only considers the number of keys in write cf. I think we should add lock cf too
> 
> @youjiali1995 yes, I wondered about that ([#7676 (comment)](https://github.com/tikv/tikv/pull/7676#issuecomment-620910114)). I don't know why we don't check the default CF too?

I guess the reason is if there is a key in default cf, there must be a same key in write cf, and not vice versa. I think the key count doesn't care about cfs.. @nrc

File: components/raftstore/src/coprocessor/split_check/size.rs [hunk 1]
```diff
@@ -267,25 +267,20 @@ fn get_approximate_split_keys(
     batch_split_limit: u64,
 ) -> Result<Vec<Vec<u8>>> {
     let get_cf_size = |cf: &str| get_region_approximate_size_cf(db, cf, &region, 0);
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply components/raftstore/src/coprocessor/split_check/size.rs:269]: emmm. return error, seems it's a problem

[the remark this answers is not in the document]
Comment (Connor1996) [reply components/raftstore/src/coprocessor/split_check/size.rs:269]: we need not return error when getting from lock cf for compatibility

**Comment by Connor1996** [comment]
> I found it only truncates timestamp for row key:
> https://github.com/tikv/tikv/blob/45201a65dc4222444681dbbc5957860b5e21c981/components/raftstore/src/coprocessor/split_observer.rs#L36-L42
> 
> Why not do the same thing for index key? Will it result in different versions of an index key in more than one regions? @Connor1996

interesting. I didn't notice it excludes the index keys. Seems the rule is from the very beginning. Do you know the reason? @BusyJay

**Comment by BusyJay** [comment]
The logic is for the old days when we still store column ID in row key, truncating is to prevent multiple column of the same row get separated. Version is truncated when decoding bytes.

File: components/raftstore/src/coprocessor/split_check/size.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (youjiali1995) [reply components/raftstore/src/coprocessor/split_check/size.rs:269]: Yes

**Comment by youjiali1995** [comment]
@nrc We should handle range properties of lock cf specially. If we fail to get the property, we assume it returns zero.

**Comment by nrc** [comment]
> We should handle range properties of lock cf specially. If we fail to get the property, we assume it returns zero.

I've addressed this, I am not confident about testing it though. I guess I can just run master TiKV, then this fork and see if it crashes?

**Comment by youjiali1995** [comment]
> > We should handle range properties of lock cf specially. If we fail to get the property, we assume it returns zero.
> 
> I've addressed this, I am not confident about testing it though. I guess I can just run master TiKV, then this fork and see if it crashes?

Yes, I think so.

**Comment by youjiali1995** [comment]
I will check and test it too.

**Comment by nrc** [comment]
> I guess I can just run master TiKV, then this fork and see if it crashes?

Seemed to work fine: no crash, no relevant error logs, nothing out of the ordinary in Grafana.

**Comment by youjiali1995** [comment]
@Connor1996 @BusyJay PTAL, thanks! We need it in v4.0.0-ga.

File: components/raftstore/src/coprocessor/split_check/size.rs [hunk 2]
```diff
@@ -266,26 +266,21 @@ fn get_approximate_split_keys(
...
-    let default_cf_size = box_try!(get_cf_size(CF_DEFAULT));
-    let write_cf_size = box_try!(get_cf_size(CF_WRITE));
-    if default_cf_size + write_cf_size == 0 {
-        return Err(box_err!("default cf and write cf is empty"));
+    let get_cf_size = |cf: &str| get_region_approximate_size_cf(db, cf, &region, 0).unwrap_or(0);
+    let cfs = [
+        (CF_DEFAULT, get_cf_size(CF_DEFAULT)),
+        (CF_WRITE, get_cf_size(CF_WRITE)),
+        (CF_LOCK, get_cf_size(CF_LOCK)),
+    ];
```
Comment (youjiali1995) [inline components/raftstore/src/coprocessor/split_check/size.rs:274]: I suggest:
```suggestion
    let get_cf_size = |cf: &str| get_region_approximate_size_cf(db, cf, &region, 0);
    let cfs = [
        (CF_DEFAULT, box_try!(get_cf_size(CF_DEFAULT))),
        (CF_WRITE, box_try!(get_cf_size(CF_WRITE))),
        (CF_LOCK, get_cf_size(CF_LOCK).unwrap_or(0)),
    ];
```
If something is wrong, the total_size will be zero forever.. I prefer not to swallow the error of CF_WRITE and CF_LOCK.

File: components/raftstore/src/coprocessor/split_check/size.rs [hunk 3]
```diff
@@ -266,26 +266,21 @@ fn get_approximate_split_keys(
...
-        )
-    } else {
-        (
-            CF_WRITE,
-            split_size * write_cf_size / (default_cf_size + write_cf_size),
-        )
-    };
+    let cf_split_size = split_size * cf_size / total_size;
 
     get_approximate_split_keys_cf(db, cf, &region, cf_split_size, max_size, batch_split_limit)
```
Comment (Connor1996) [inline components/raftstore/src/coprocessor/split_check/size.rs:285]: `get_approximate_split_keys_cf` will return error in L303 for lock cf

Comment (youjiali1995) [reply components/raftstore/src/coprocessor/split_check/size.rs:285]: If `get_region_approximate_size_cf`doesn't report error, I think  `get_approximate_split_keys_cf` won't report error too.

**Comment by youjiali1995** [comment]
Sorry, @nrc . I pushed some commits because I think you are absent. 😅

**Comment by youjiali1995** [comment]
> How about adding a test to make sure everything is okay from upgrading a old version instance. I'm afraid of missing some other places

I'm trying.

**Comment by youjiali1995** [comment]
> > How about adding a test to make sure everything is okay from upgrading a old version instance. I'm afraid of missing some other places
> 
> I'm trying.

I failed. I'm on the train...

**Comment by youjiali1995** [comment]
@Connor1996 PTAL, thanks!

**Pull request merged**